ecosystem/nft-metadata-crawler-parser/src/parser.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    config::ParserConfig,
+    utils::{
+        counters::{
+            GOT_CONNECTION_COUNT, PARSER_FAIL_COUNT, PARSER_INVOCATIONS_COUNT,
+            PUBSUB_ACK_SUCCESS_COUNT, SKIP_URI_COUNT, UNABLE_TO_GET_CONNECTION_COUNT,
+        },
+        database::check_or_update_chain_id,
+    },
+    worker::Worker,
+    Server,
+};
+use axum::response::Response;
+use bytes::Bytes;
+use diesel::{
+    r2d2::{ConnectionManager, Pool},
+    PgConnection,
+};
+use google_cloud_storage::client::{Client as GCSClient, ClientConfig as GCSClientConfig};
+use reqwest::StatusCode;
+use std::sync::Arc;
+use tracing::{error, info, warn};
+
+/// Struct to hold context required for parsing
+#[derive(Clone)]
+pub struct ParserContext {
+    pub parser_config: Arc<ParserConfig>,
+    pub pool: Pool<ConnectionManager<PgConnection>>,
+    pub gcs_client: Arc<GCSClient>,
+    pub max_num_retries: i32,
+}
+
+impl ParserContext {
+    pub async fn new(
+        parser_config: ParserConfig,
+        pool: Pool<ConnectionManager<PgConnection>>,
+        max_num_retries: i32,
+    ) -> Self {
+        if let Some(google_application_credentials) = &parser_config.google_application_credentials
+        {
+            info!(
+                "[NFT Metadata Crawler] Google Application Credentials path found, setting env var"
+            );
+            std::env::set_var(
+                "GOOGLE_APPLICATION_CREDENTIALS",
+                google_application_credentials,
+            );
+        }
+
+        // Establish GCS client
+        let gcs_config = GCSClientConfig::default()
+            .with_auth()
+            .await
+            .unwrap_or_else(|e| {
+                error!(
+                    error = ?e,
+                    "[NFT Metadata Crawler] Failed to create gRPC client config"
+                );
+                panic!();
```

Comment:
Make this return a Result and error normally rather than panicking?

Reply:
I think it's ok to just panic here bc this is required for setting up the connection to GCS and is run at the very beginning